src/graph/network/msg_queue.cc

```diff
@@ -0,0 +1,189 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file msg_queue.cc
+ * \brief Message queue for DGL distributed training.
+ */
+#include <dmlc/logging.h>
+#include <cstring>
+
+#include "msg_queue.h"
+
+namespace dgl {
+namespace network {
+
+using std::string;
+
+MessageQueue::MessageQueue(int queue_size, int num_producers) {
+  CHECK_LT(0, queue_size);
+  try {
+    queue_ = new char[queue_size];
+  } catch(const std::bad_alloc&) {
+    LOG(FATAL) << "Not enough memory for message buffer.";
+  }
+  memset(queue_, '\0', queue_size);
+
+  queue_size_ = queue_size;
+  free_size_ = queue_size;
+  write_pointer_ = 0;
+  num_producers_ = num_producers;
+}
+
+MessageQueue::~MessageQueue() {
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (nullptr != queue_) {
+    delete [] queue_;
+    queue_ = nullptr;
+  }
+}
+
+int MessageQueue::Add(const char* src, int size, bool is_blocking) {
```

Comment:
this is a large memory copy.

Reply:
It's not very easy to implement zero-copy here

Reply:
how much does the memory copy affect the performance?

Reply:
Let me evaluate it by removing the message queue for a test.